### ISSUE TITLE
Extruded drawtypes

### ISSUE THREAD

**Issue by MirceaKitsune** [minetest/minetest#5666]
I wanted to bring attention to a feature that seems to have been forgotten, which I'd really like to see revived under some form. It was one of the things RealBadAngel had been working on, before the unfortunate event of his death occurred. Alongside shaders such as Bloom or Depth-of-Field, it was one of the visual improvements I was waiting most eagerly to see finished.

The feature allows for all flat drawtypes, such as "torch" or "plantlike", to use the extrusion system of the wielded item and have a width. This results in a much more beautiful and artistic appearance for plants and other items, as seen in the screenshots posted at the time:

![screenshot1](https://cloud.githubusercontent.com/assets/2177790/5161478/3b8e1d06-73aa-11e4-9f76-81ce41f81347.png)
![screenshot2](https://cloud.githubusercontent.com/assets/2177790/12025577/a361cde2-adb0-11e5-84e5-9fd8b4ecf0f8.png)
![screenshot3](https://cloud.githubusercontent.com/assets/2177790/12025676/8773c1de-adb1-11e5-8e42-72da37c81265.png)

Unfortunately both the original issue and pull request have been closed down and abandoned. They can still be found here, along with the code and discussions:

#1871
#3496

From what I understand, the problem was that this required new drawtype definitions, and it was suggested that a parameter to define the width which works for all drawtypes is used instead. I perfectly agree with this myself... sadly RBA was no longer around to finish up on the change.

Would it be possible for someone else to please pick up this code, and finalize the last concerns expressed about it? It is a beautiful improvement especially for grass and flowers, and I really wish to see it as part of the engine and use it at some point! Thank you.

**Comment by paramat** [comment]
:-1: See my arguments in #3496
See https://github.com/minetest/minetest/pull/3496#issuecomment-167742215 this was closed by celeron55 as being unnacceptable (rightly so too).

**Comment by paramat** [comment]
Discussion should continue in #3496, this is duplicate.

**Comment by MirceaKitsune** [comment]
@paramat Ah, I might have misunderstood his argument. In that case I can't say I agree with it, and hope this can be reconsidered and discussed again without getting closed too.

Although something similar may be achieved using mesh nodes, I believe this should still exist as a separate feature for 2D drawtypes. It works by extruding the texture, meaning it automatically adjusts to any image offering an important bit of flexibility! If it was a model, and someone wanted to update the texture, a whole new model would have to be created instead.

Also notice how each pixel is separated into a little cube and randomly offset, something that would be more difficult to do manually and require storing a high polyon model. The waving might also not work with mesh... it would require a strange form of vertex animation, and even then the engine doesn't support animated meshnodes IIRC. Other visual and performance optimizations might also be possible to do using this system.

**Comment by MirceaKitsune** [comment]
Um... the old issue is indeed a duplicate, but it was closed long ago. Basically you're closing this because the other issue was closed too! Could either of them be reopened at least?

**Comment by paramat** [comment]
:-1: See my arguments in #3496
> Also notice how each pixel is separated into a little cube and randomly offset,

![extruded_masses](https://cloud.githubusercontent.com/assets/3192173/6097184/5865f5d0-afb3-11e4-951e-1367fe3f62f8.png)

With the amount of plantlike nodes we have the millions of extra cubes to render will be a disaster for performance, so this will need to be avoided.

**Comment by MirceaKitsune** [comment]
If the individual pixel cubes are as slow, I agree. Although they could still be made into a separate setting, so mod / game developers can enable them on rarer items where they wouldn't have a considerable performance impact. 0 would disable it, > 0 would adjust the randomness and magnitude of the cubes possibly. Personally I'm not as picky about this detail, though it looks a bit too good to throw out if we could keep it separately.

I believe one important argument is left in favor of implementing drawtype extrusion: It automatically generates the model for every texture, accounting for its alpha channel. If a texture update changes the transparency, a fixed model would also require manual updating. Also everyone making a texture package would have to make custom models for every extruded drawtype. Not every texture artist might have the tools or knowledge needed to make models! So for something as common and basic such as simple extrusion, an automated way of generating the mesh would be an important convenience.

**Comment by paramat** [comment]
Supporting a 2nd parallel implementation for a rare use of pixel cubes is not a good idea, especially since such usage could only be used for a very few nodes.

From IRC:
Paramat:
so anyway the individual pixel cubes are not doable, but a continuous waving mesh is
looks better too
best would be to automatically extrude a texture and convert it into the existing meshnode format, as nodeboxes are
so a completely new implementation is needed

**Comment by MirceaKitsune** [comment]
The little cubes can go as far as I care. Normal mapping compensates for that effect anyway.

I'm mainly interested in at least simple extrusion, like that of the wielded item. Which itself would be a widespread use case, especially for plants on which it looks very beautiful.

**Comment by paramat** [comment]
WIth a continuous meshnode ther are still rendering issues.
For example a 'grass 5' node has 2*2*76 = 304 vertices. Plantlike nodes are inherently used in large numbers with often thousands in view.
However the small extrusion would only be visible for the few plantlike nodes close to the player, beyond about 8 nodes it looks just like a flat plane. So the overwhelming majority of plantlikes in view will be extruded for no reason, but that majority would also cause a very heavy load on rendering.
So i don't think it's a good idea to work on this in MT, for those masochists who want to do this maybe a separate program can be written to convert any texture into a meshnode.

**Comment by paramat** [comment]
More personally and subjectively, i find extruded textures look much worse than flat textures. A flat texture is just being itself in its ultra-simple abstract way. When that is extruded it becomes a flat texture trying and completely failing to be 3D, it becomes artistically false and distasteful, while looking barely any different.
If we have extra-high resolution small flora that should be done with freshly-designed dedicated 3D models, and we can already do that with nodeboxes and meshes.
So sorry but i still consider this one of the worst ideas ever suggested for MT.

**Comment by Calinou** [comment]
> When that is extruded it becomes a flat texture trying and completely failing to be 3D, it becomes artistically false and distasteful, while looking barely any different.

I agree with this, it also tends to make the node itself harder to identify, especially between close variations.

**Comment by sofar** [comment]
I don't think this looks bad, but I'm not a fan of the complexity it creates. Even at a distance of 8 nodes, we're talking about a potential of 843 million triangles if someone thinks 32px textures are a cool idea.

whoops.

Someone calculate how many tris are in a single grass plantlike plant with HDX.

**Comment by kilbith** [comment]
:-1: Please close this stupid issue.

512^2 pixels texture = 262,144 pixels in total.
Let's say the texture has 60% of colored pixels, the rest is alpha and non-extruded : 157,286 pixels.
Each pixel is converted to a cube (= 8 vertices) = 1,258,288 vertices, per node.

Now let's say the texture is a tall grass spread over a land, we have e.g. 100 nodes of viewing range, and 60% of the area is covered of grass = 4716 nodes * 1,258,288 vertices = 5,934,086,208 vertices. Needless to say, this is insanely high.

**Comment by MirceaKitsune** [comment]
@kilbith The fact that you don't like it doesn't make it stupid. Myself, RealBadAngel (who initially coded it), and others, find it visually appealing. I don't mind if like many effects, it's disabled by default... however it would be a very nice option for us to have in the engine.

If this is considered unnecessary complexity, the builtin system to extrude the wielded item (or 3D clouds) might just as well qualify as such; Why not either let the wield item be flat, or require a mesh for every extruded item? And by the way, I'm basically proposing a way of using the same algorithm as the wield item on the "plantlike" drawtype... so it's more like piping something that already exists into another use case, rather than coding in something 100% different for no reason.

Preserving performance and not abusing this feature would in part be the responsibility of the game, namely the maintainer of the mapgen configuration. Custom biomes may add more or less grass, whereas this can be used only for rarer types of plants (such as flowers).

To further improve on this, it could come with a distance based feature (similarly to LOD): Past a customizable view distance, the effect gets disabled and nodes are rendered using the original flat surfaces. This is yet another thing you can't do as cleanly with meshes.

**Comment by tobyplowy** [comment]
@MirceaKitsune I think this is a cool feature and it looks really good but I don't think it's worth the performance hit!

Maybe something can be done so it only renders a certain distance like 20 nodes max?
But apart from that I don't think this is a good idea :/

**Comment by MirceaKitsune** [comment]
@tobyplowy That's what I was suggesting, and I think it's a great idea: The setting represents the distance at which the effect shows... either as a fixed distance in nodes, or a percentage between the min and max view distance. The effect is only obvious from up close, so something like 10 to 20 nodes would be a good default.

**Comment by stujones11** [comment]
I personally quite like the look of them, perhaps written as an optional (off by default) shader the efficiency could be improved but in its current form it is simply not worth the performance hit on even the best hardware.

**Comment by dsohler** [comment]
> Now let's say the texture is a tall grass spread over a land, we have e.g. 100 nodes of viewing range, and 60% of the area is covered of grass = 4716 nodes * 1,258,288 vertices = 5,934,086,208 vertices. Needless to say, this is insanely high.

Now have the render engine remove all parts that are not visible to the player. Boom, down to a few hundred at max.

**Comment by MirceaKitsune** [comment]
@doserj I consider there's a very good reason for this feature: It automatically generates the 3D effect based on the texture and its alpha channel on the client side. This is times more practical over doing it with fixed 3D models, for a variety of reasons that I can think of:

1. If it's a fixed model, you need to update it every time you update its texture (in case you're also updating its alpha channel which is almost always the case).
2. If it's a fixed model, every custom texture package needs to make new models for every texture belonging to an extruded item (again if it changes the alpha channel).
3. Not every texture artist has the tools and knowledge to create models. Someone might be able to create a texture in Photoshop, but not know Blender to make an extruded mesh for it.
4. This effect is extremely tedious to model manually! Imagine opening up a 3D modeling program, placing little cubes one next to another, UV mapping each to match a precise pixel on an image, then removing hidden neighboring surfaces to get rid of extra vertices... it's a nightmare.
5. As you need to distribute the model alongside the texture, clients connecting to servers might need to download it upon joining. In an engine where even 16px textures are optimized to the bone to reduce file size, countless 3D models for such a simple effect would likely not be desired.
6. The exact same effect already exists in the engine and is widely used, for both wielded and dropped items as well as 3D clouds. Therefore this isn't even suggesting a completely new feature coded from scratch, just piping an existing algorithm into a few other use cases.
7. Because we're doing the same thing with items, the effect also fits in from an artistic standpoint... not only does it not look out of place, but it's in fact the most logical result: The wielded item is normally flat too, and we extrude it to make it prettier... plants and torches are in the same situation, so why would they not be extruded? Think back to the old flat torches, or consider saplings for instance: If you hold one in your hand, it appears extruded... but once you placed it on the ground, the extrusion effect vanishes... logically this means the item changes shape which doesn't make much sense.

I think every point on this list is a good reason to consider the functionality. I agree with it staying disabled by default, if more people think it's ugly than they think it's pretty. However minetest_game should make it available by default, for every client that turns on this setting. Perhaps a single option should control extrusion as follows: 0 = disabled, 1 = wielditem and dropped items (current behavior, remains default), 2 = wielditem and dropped items + flat drawtypes.

Once more, I fully agree with a simple LOD system which should fix all performance concerns, where the effect is simply disabled for nodes further than a customizable view distance. Also keep in mind that with simple flat extrusion, vertices are only generated for the edges of pixels neighboring transparent pixels, therefore it's not a cube for each pixel and vertice count depends strictly on the complexity of the texture's alpha channel.

**Comment by pithydon** [comment]
In reason 4 there is a better way to do that but it does not matter because of reason 3.

**Comment by raymoo** [comment]
Seems like something that could be done efficiently with geometry shaders.

**Comment by paramat** [comment]
2 disapprovals and no core dev support.

**Comment by MirceaKitsune** [comment]
How easy it would have been to add such a feature, for such a beautiful effect too. It doesn't feel like Minetest is going to have developers interested in the visual aspect again sadly. I can hope that someday at least a fork may continue the development and focus that stopped when RealBadAngel left.

**Comment by rubenwardy** [comment]
A fork is unlikely to happen due to the same reason minetest is stagnating on graphics - we don't have anyone who knows graphics well

**Comment by paramat** [comment]
Ease of adding the feature is irrelevant considering the obvious performance problems.
All MT devs are interested in the visual aspect, obviously. Not being interested in fancy visual effects does not equate to lack of interest in visuals, that's a little insulting. Maybe you should find a new game instead of being negative here? It seems MT is not to your taste.
We do have contributors interested in working on shaders and post-processing.

RBA didn't create as much development and focus as you imply, and he wasn't as talented as you imply, his code was usually rather sloppy and problematic. We only have a few shaders, some of those not implemented well, in my opinion ugly, and development was slow.

**Comment by MirceaKitsune** [comment]
I got an email notification that this was updated since I created the ticket, that's why I "popped up". If MineTest wasn't to my liking, I wouldn't have a reason to be upset about any choices it does (or doesn't) take because I wouldn't care... if I'm upset it's because I do care. The base structure and functionality at the core of the engine is wonderfully done, I just feel that modernization in other areas (lighting, shaders, etc) simply didn't have the chance to happen... this is frustrating because you feel that something good could be even better but doesn't take that opportunity, and abandoning this suggestion made me feel that yet again. I don't mean to trigger an off-topic debate here, though discussion about the stagnation in graphics development has been around many times now.
